File: test/lib/vbrain/links_test.rb
require "test_helper"
require "vbrain/links"

class LinksTest < Minitest::Test
  def test_extracts_single_link
    assert_equal ["Foo Bar"], VBrain::Links.extract("texto com [[Foo Bar]] no meio")
  end

  def test_extracts_multiple_in_order
    body = "Veja [[Alpha]] e depois [[Beta]] e [[Gamma]]."
    assert_equal %w[Alpha Beta Gamma], VBrain::Links.extract(body)
  end

  def test_supports_alias_keeps_target_only
    assert_equal ["Postgres Logical"], VBrain::Links.extract("ver [[Postgres Logical|replicação]]")
  end

  def test_dedups_repeated_targets
    assert_equal ["X"], VBrain::Links.extract("[[X]] e de novo [[X]] e [[X|outro texto]]")
  end

  def test_strips_whitespace
    assert_equal ["Espaçado"], VBrain::Links.extract("[[  Espaçado  ]]")
  end

  def test_returns_empty_for_body_without_links
    assert_equal [], VBrain::Links.extract("sem nenhum link aqui [x] (y) {z}")
  end

  def test_ignores_empty_or_whitespace_only_links
    assert_equal [], VBrain::Links.extract("[[]] e [[   ]] e [[|só alias]]")
  end

  def test_returns_empty_for_nil
    assert_equal [], VBrain::Links.extract(nil)
  end

  def test_target_slug_matches_write_pages_slug
    # Mesma normalização que VBrain::Slug.from, pra resolução casar.
    assert_equal "foo-bar", VBrain::Links.target_slug("Foo Bar")
    assert_equal "replicacao-logica", VBrain::Links.target_slug("Replicação Lógica")
  end

  def test_target_slug_returns_nil_when_no_valid_slug
    assert_nil VBrain::Links.target_slug("!!!")
    assert_nil VBrain::Links.target_slug("   ")
  end
end

File: internal/links/links_test.go
package links

import (
	"reflect"
	"testing"
)

func slugsOf(body string) []string {
	var out []string
	for _, l := range Extract(body) {
		out = append(out, l.Slug)
	}
	return out
}

func TestExtractsSingleWikilink(t *testing.T) {
	got := Extract("texto com [[Foo Bar]] no meio")
	if len(got) != 1 || got[0].Slug != "foo-bar" || got[0].Title != "Foo Bar" {
		t.Fatalf("got %+v", got)
	}
}

func TestExtractsMultipleInOrder(t *testing.T) {
	if got := slugsOf("[[Alpha]] e [[Beta]] e [[Gamma]]."); !reflect.DeepEqual(got, []string{"alpha", "beta", "gamma"}) {
		t.Fatalf("got %v", got)
	}
}

func TestWikilinkAliasKeepsTargetForSlugAndTitle(t *testing.T) {
	got := Extract("ver [[Postgres Logical|replicação]]")
	if got[0].Slug != "postgres-logical" || got[0].Title != "Postgres Logical" {
		t.Fatalf("got %+v", got[0])
	}
}

func TestDedupsBySlug(t *testing.T) {
	if got := slugsOf("[[X]] e de novo [[X]] e [[X|outro]]"); !reflect.DeepEqual(got, []string{"x"}) {
		t.Fatalf("got %v", got)
	}
}

func TestParsesMarkdownLinksToMdFiles(t *testing.T) {
	got := Extract("veja [Família de Victor](familia-de-victor.md) aqui")
	if got[0].Slug != "familia-de-victor" || got[0].Title != "Família de Victor" {
		t.Fatalf("got %+v", got[0])
	}
}

func TestBothFormsDedupBySlug(t *testing.T) {
	body := "[[Família de Victor]] e depois [Família de Victor](familia-de-victor.md)"
	if got := slugsOf(body); !reflect.DeepEqual(got, []string{"familia-de-victor"}) {
		t.Fatalf("got %v", got)
	}
}

func TestIgnoresExternalAndNonMdMarkdownLinks(t *testing.T) {
	if got := slugsOf("[Google](https://google.com) e [foto](pic.png)"); len(got) != 0 {
		t.Fatalf("got %v", got)
	}
}

func TestEmptyAndNil(t *testing.T) {
	for _, in := range []string{"sem link [x] (y)", "", "[[]] e [[   ]]"} {
		if got := Extract(in); len(got) != 0 {
			t.Errorf("Extract(%q) = %v, want empty", in, got)
		}
	}
}

func TestLinkifyConvertsResolvableWikilink(t *testing.T) {
	out := Linkify("liga pra [[Família de Victor]].", []string{"familia-de-victor"})
	if out != "liga pra [Família de Victor](familia-de-victor.md)." {
		t.Fatalf("got %q", out)
	}
}

func TestLinkifyLeavesUnresolvableUntouched(t *testing.T) {
	body := "liga pra [[Página Inexistente]]."
	if out := Linkify(body, []string{"familia-de-victor"}); out != body {
		t.Fatalf("got %q", out)
	}
}

func TestLinkifyAliasUsesDisplayTextAndTargetSlug(t *testing.T) {
	out := Linkify("ver [[Postgres Logical|replicação lógica]]", []string{"postgres-logical"})
	if out != "ver [replicação lógica](postgres-logical.md)" {
		t.Fatalf("got %q", out)
	}
}

func TestLinkifyIsIdempotent(t *testing.T) {
	body := "liga pra [[Família de Victor]]."
	once := Linkify(body, []string{"familia-de-victor"})
	twice := Linkify(once, []string{"familia-de-victor"})
	if once != twice {
		t.Fatalf("not idempotent: %q != %q", once, twice)
	}
}

func TestLinkifyAcceptsArrayOfSlugs(t *testing.T) {
	if out := Linkify("[[A]]", []string{"a"}); out != "[A](a.md)" {
		t.Fatalf("got %q", out)
	}
}

func TestApplyResolutionMapsTitleToChosenSlug(t *testing.T) {
	out := ApplyResolution("trabalha na [[V360]].", map[string]string{"V360": "carreira-de-victor"})
	if out != "trabalha na [V360](carreira-de-victor.md)." {
		t.Fatalf("got %q", out)
	}
}

func TestApplyResolutionAliasUsesDisplay(t *testing.T) {
	out := ApplyResolution("ver [[Empresa V360|a V360]]", map[string]string{"Empresa V360": "v360"})
	if out != "ver [a V360](v360.md)" {
		t.Fatalf("got %q", out)
	}
}

func TestApplyResolutionLeavesNullOrAbsentUntouched(t *testing.T) {
	body := "[[UFRJ]] e [[Outra]]"
	// slug "" (LLM found nothing) and a title absent from the map stay intact.
	if out := ApplyResolution(body, map[string]string{"UFRJ": ""}); out != body {
		t.Fatalf("got %q", out)
	}
}

func TestApplyResolutionEmptyMapNoop(t *testing.T) {
	body := "[[X]]"
	if out := ApplyResolution(body, map[string]string{}); out != body {
		t.Fatalf("got %q", out)
	}
}
